### PR TITLE
HEEDLS-585 All delegates - welcome emails - filter - fix

### PR DESCRIPTION
**Pull request by ibrahimmunir14**
### JIRA link
[HEEDLS-585](https://softwiretech.atlassian.net/browse/HEEDLS-585)

### Description
- Fix JS bug: If I add a filter, and then go to add a different filter, but leave it on the default value of “Select [filter name]” and hit the Add Filter button then nothing happens as expected. However, I then cannot add any new filters until I clear all the filters.
- Word break for really long delegate checkbox labels
- Read out results count when selecting/deselect all

### Screenshots
Word wrap:

-----
### Developer checks
I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (controller, data services, services, view models etc) and manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme.
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Web/Styles/trackingSystem/delegates/emailDelegates.scss [hunk 1]
```diff
@@ -5,3 +5,7 @@
   border-radius: 4px;
   margin-right: nhsuk-spacing(2);
 }
+
+.delegate-label {
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Styles/trackingSystem/delegates/emailDelegates.scss:9]: This fix won't work as expected in IE, use the word-break class on the elements (This is defined in index.scss)

File: DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts [hunk 2]
```diff
@@ -8,18 +8,26 @@ const route = `TrackingSystem/Delegates/Email/AllEmailDelegateItems?${queryParam
...
+  resultCount.setAttribute('role', '');
+  resultCount.setAttribute('role', 'alert');
+}
+
 function selectAll(): void {
   const allCheckboxes = document.querySelectorAll('.delegate-checkbox') as NodeListOf<HTMLInputElement>;
   allCheckboxes.forEach((checkbox) => {
     if (!checkbox.checked) checkbox.click();
   });
+  alertResultCount();
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts:22]: I think these alert calls need to go in the event listener declaration. It's not a part of the select all, it belongs to the button press i.e. the button action becomes "Select all then alert the user"
I've moved the selectAll functions to a common place as part of 441, and not all pages will necessarily have the results count element to be able to do this and will need an alternative way to announce it if we want this functionality.

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
A couple of minor things that need modifying.

**Review by AlexJacksonDS — APPROVED** [review]

File: DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts [hunk 3]
```diff
@@ -27,5 +33,11 @@ const selectAllButton = document.getElementById('select-all-button') as HTMLButt
 const deselectAllButton = document.getElementById('deselect-all-button') as HTMLButtonElement;
 
 selectAllForm.addEventListener('submit', (e) => e.preventDefault());
-selectAllButton.addEventListener('click', selectAll);
-deselectAllButton.addEventListener('click', deselectAll);
+selectAllButton.addEventListener('click', () => {
+  selectAll();
+  alertResultCount();
+});
+deselectAllButton.addEventListener('click', () => {
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts:40]: Minor, but missing an empty line above.

Also, can we wrap all this with setUpSelectAndDeselectButtons() function and call that at the top of the page (so devs don't have to scroll through the whole file to see what is happening on page load).

File: DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts [hunk 4]
```diff
@@ -8,6 +8,12 @@ const route = `TrackingSystem/Delegates/Email/AllEmailDelegateItems?${queryParam
 // eslint-disable-next-line no-new
 new SearchSortFilterAndPaginate(route, false, false, true, 'EmailDelegateFilter');
 
+function alertResultCount(): void {
+  const resultCount = document.getElementById('results-count') as HTMLSpanElement;
+  resultCount.setAttribute('role', '');
+  resultCount.setAttribute('role', 'alert');
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts:14]: What does this do? Maybe worth a comment if this is a hack to let screen reader announce the message/or rename to announceNewResultCount()?
Alert usually means popup on a page in terms of JS (try the alert method in console):
image

Comment (ibrahimmunir14) [reply DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts:14]: If a html element has role=alert it gets read by the screen reader. In the filtering stuff, whenever a filter is changed, the text on the element changes which makes it re-read. Here the text isn't changing, so we can just unassign and reassign the role alert to make it re-read.

File: DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts [hunk 5]
```diff
@@ -7,8 +7,10 @@ const route = `TrackingSystem/Delegates/Email/AllEmailDelegateItems?${queryParam
 
 // eslint-disable-next-line no-new
 new SearchSortFilterAndPaginate(route, false, false, true, 'EmailDelegateFilter');
+setupSelectAndDeselectButtons();
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Scripts/trackingSystem/emailDelegates.ts:10]: Super minor, but given this is a verb, it should technically be "setUp" re capitalisation 👁️

**Review by stellake — APPROVED** [review]
Super minor comment about capitalisation, but no need for rereview

**Pull request merged**